SSA/Experimental/Bits/Fast/BitStream.lean

```diff
@@ -98,6 +98,15 @@ section Lemmas
 theorem ext {x y : BitStream} (h : ∀ i, x i = y i) : x = y := by
   funext i; exact h i
 
+theorem corec_eq_corec {a : α} {b : β} {f g}
+    (R : α → β → Prop)
+    (h : ∀ a b, R a b →
+          let x := f a
+          let y := g b
+          R x.fst y.fst ∧ x.snd = y.snd) :
+    corec f a = corec g b := by
+  sorry
```

Comment:
@alexkeizer I don't understand the purpose of this theorem? And is it even true?

I feel like the correct version should be 

```lean

theorem corec_eq_corec {a : α} {b : β} {f g}
    (R : α → β → Prop)
    (thing : R a b)
    (h : ∀ a b, R a b →
          let x := f a
          let y := g b
          R x.fst y.fst ∧ x.snd = y.snd) :
    corec f a = corec g b := by
  ext i
  induction' i with i ih 
  simp [corec]
  have y := h a b thing
  simp at y
  cases' y with l r
  assumption
  sorry
```

Reply:
I filled out the proof as this:


```lean

theorem compose_first {α: Type u₁} (i : Nat) (a : α ) (f : α → α × Bool) : (f ((Prod.fst ∘ f)^[i] a)).1 = (Prod.fst ∘ f)^[i] (f a).1 :=
  match i with
    | 0 => by simp
    | i + 1=>
      let y:= @compose_first α i ((f a).1) f
      by simp [y]

theorem corec_eq_corec {a : α} {b : β} {f g}
    (R : α → β → Prop)
    (thing : R a b)
    (h : ∀ a b, R a b →
          let x := f a
          let y := g b
          R x.fst y.fst ∧ x.snd = y.snd) :
    corec f a = corec g b := by
  ext i
  have lem :  R ((Prod.fst ∘ f)^[i] (f a).1) ((Prod.fst ∘ g)^[i] (g b).1) ∧ corec f a i = corec g b i := by
    induction' i with i ih
    <;> simp [corec]
    apply h
    exact thing
    simp at h
    simp [corec] at ih
    have m := h ((Prod.fst ∘ f)^[i] (f a).1) ((Prod.fst ∘ g)^[i] (g b).1) (ih.1)
    cases' m with l r
    simp [r]
    have left : ((Prod.fst ∘ f)^[i] (f (f a).1).1)  = (f ((Prod.fst ∘ f)^[i] (f a).1)).1 := by
      rw [← @compose_first α]
    have right : ((Prod.fst ∘ g)^[i] (g (g b).1).1)  =(g ((Prod.fst ∘ g)^[i] (g b).1)).1  := by
      rw [← @compose_first β i]
    simp [left, right]
    exact l
  cases' lem with l r
  assumption
```